File: api/v1/schemas/trading.py
# -*- coding: utf-8 -*-
"""Pydantic schemas for the trading API.

Snake-case fields (the frontend's ``toCamelCase`` helper handles the
boundary). All numeric / string types match the underlying dataclass
shapes in ``src/trading/types.py``.
"""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class TradingStatusResponse(BaseModel):
    status: str  # disabled | ready | error
    mode: str    # disabled | paper | live
    message: Optional[str] = None
    paper_account_id: Optional[int] = None
    max_position_value: Optional[float] = None
    max_position_pct: Optional[float] = None
    max_daily_turnover: Optional[float] = None
    symbol_allowlist: List[str] = Field(default_factory=list)
    symbol_denylist: List[str] = Field(default_factory=list)
    market_hours_strict: Optional[bool] = None
    notification_enabled: Optional[bool] = None


class OrderSubmitRequest(BaseModel):
    symbol: str = Field(..., min_length=1, max_length=16)
    side: Literal["buy", "sell"]
    quantity: float = Field(..., gt=0)
    order_type: Literal["market", "limit"] = "market"
    limit_price: Optional[float] = Field(None, gt=0)
    time_in_force: Literal["day", "gtc"] = "day"
    account_id: Optional[int] = None
    market: Optional[Literal["us", "cn", "hk"]] = None
    currency: Optional[str] = Field(None, min_length=3, max_length=8)
    note: Optional[str] = Field(None, max_length=255)
    request_uid: str = Field(..., min_length=8, max_length=64)
    source: Literal["ui", "agent", "strategy"] = "ui"
    agent_session_id: Optional[str] = Field(None, max_length=100)


class RiskFlagItem(BaseModel):
    code: str
    severity: str
    message: str
    detail: Dict[str, Any] = Field(default_factory=dict)


class RiskAssessmentItem(BaseModel):
    flags: List[RiskFlagItem] = Field(default_factory=list)
    decision: str
    evaluated_at: str
    config_snapshot: Dict[str, Any] = Field(default_factory=dict)


class OrderResultItem(BaseModel):
    request: Dict[str, Any]
    status: str
    mode: str
    fill_price: Optional[float] = None
    fill_quantity: Optional[float] = None
    fill_time: Optional[str] = None
    realised_fee: float = 0.0
    realised_tax: float = 0.0
    risk_assessment: Optional[RiskAssessmentItem] = None
    portfolio_trade_id: Optional[int] = None
    error_code: Optional[str] = None
    error_message: Optional[str] = None
    quote_payload: Optional[Dict[str, Any]] = None


class OrderSubmitResponse(OrderResultItem):
    pass


class RiskPreviewRequest(OrderSubmitRequest):
    """Same shape as a submit request — preview just runs the
    RiskEngine and returns the assessment without writing audit."""


class RiskPreviewResponse(BaseModel):
    decision: str
    flags: List[RiskFlagItem] = Field(default_factory=list)
    evaluated_at: str
    config_snapshot: Dict[str, Any] = Field(default_factory=dict)


class TradeExecutionItem(BaseModel):
    id: int
    request_uid: str
    mode: str
    source: str
    symbol: str
    side: str
    order_type: str
    quantity: float
    limit_price: Optional[float] = None
    account_id: Optional[int] = None
    market: Optional[str] = None
    currency: Optional[str] = None
    status: str
    risk_decision: Optional[str] = None
    risk_flags: List[RiskFlagItem] = Field(default_factory=list)
    fill_price: Optional[float] = None
    fill_quantity: Optional[float] = None
    realised_fee: float = 0.0
    realised_tax: float = 0.0
    portfolio_trade_id: Optional[int] = None
    error_code: Optional[str] = None
    error_message: Optional[str] = None
    agent_session_id: Optional[str] = None
    requested_at: Optional[str] = None
    finished_at: Optional[str] = None
    created_at: Optional[str] = None
    request_payload: Dict[str, Any] = Field(default_factory=dict)
    result_payload: Optional[Dict[str, Any]] = None


class TradeExecutionListResponse(BaseModel):
    items: List[TradeExecutionItem] = Field(default_factory=list)
    count: int = 0

File: apps/dsa-web/src/types/trading.ts
/**
 * TypeScript types for the trading framework Phase A
 * (`/api/v1/trading/*`).
 *
 * Mirrors `api/v1/schemas/trading.py` after `toCamelCase` on the
 * frontend boundary. Phase A is paper-only; live mode is wired but
 * stubs out at the executor — endpoints return 503 with
 * `error_code='LIVE_NOT_IMPLEMENTED'` if `TRADING_MODE=live`.
 */

export type TradingMode = 'disabled' | 'paper' | 'live';
export type OrderSide = 'buy' | 'sell';
export type OrderTypeKind = 'market' | 'limit';
export type TimeInForceKind = 'day' | 'gtc';
export type ExecutionStatusKind =
  | 'pending'
  | 'filled'
  | 'blocked'
  | 'failed';
export type RiskSeverityKind = 'info' | 'warning' | 'block';

export interface TradingStatus {
  status: 'disabled' | 'ready' | string;
  mode: TradingMode | string;
  message?: string | null;
  paperAccountId?: number | null;
  maxPositionValue?: number | null;
  maxPositionPct?: number | null;
  maxDailyTurnover?: number | null;
  symbolAllowlist: string[];
  symbolDenylist: string[];
  marketHoursStrict?: boolean | null;
  notificationEnabled?: boolean | null;
}

export interface OrderSubmitPayload {
  symbol: string;
  side: OrderSide;
  quantity: number;
  orderType: OrderTypeKind;
  limitPrice?: number | null;
  timeInForce?: TimeInForceKind;
  accountId?: number | null;
  market?: 'us' | 'cn' | 'hk' | null;
  currency?: string | null;
  note?: string | null;
  requestUid: string;
  source?: 'ui' | 'agent' | 'strategy';
  agentSessionId?: string | null;
}

export interface RiskFlag {
  code: string;
  severity: RiskSeverityKind | string;
  message: string;
  detail?: Record<string, unknown>;
}

export interface RiskAssessment {
  decision: 'allow' | 'block' | string;
  flags: RiskFlag[];
  evaluatedAt: string;
  configSnapshot: Record<string, unknown>;
}

export interface OrderResult {
  request: Record<string, unknown>;
  status: ExecutionStatusKind | string;
  mode: TradingMode | string;
  fillPrice?: number | null;
  fillQuantity?: number | null;
  fillTime?: string | null;
  realisedFee?: number;
  realisedTax?: number;
  riskAssessment?: RiskAssessment | null;
  portfolioTradeId?: number | null;
  errorCode?: string | null;
  errorMessage?: string | null;
  quotePayload?: Record<string, unknown> | null;
}

export interface TradeExecution {
  id: number;
  requestUid: string;
  mode: string;
  source: string;
  symbol: string;
  side: string;
  orderType: string;
  quantity: number;
  limitPrice?: number | null;
  accountId?: number | null;
  market?: string | null;
  currency?: string | null;
  status: string;
  riskDecision?: string | null;
  riskFlags: RiskFlag[];
  fillPrice?: number | null;
  fillQuantity?: number | null;
  realisedFee?: number;
  realisedTax?: number;
  portfolioTradeId?: number | null;
  errorCode?: string | null;
  errorMessage?: string | null;
  agentSessionId?: string | null;
  requestedAt?: string | null;
  finishedAt?: string | null;
  createdAt?: string | null;
  requestPayload: Record<string, unknown>;
  resultPayload?: Record<string, unknown> | null;
}

export interface TradeExecutionList {
  items: TradeExecution[];
  count: number;
}
